### PR TITLE
Subscriptions

### PR DESCRIPTION
**Pull request by MStokluska**
### Summary

Added subscriptions. 

### To test

1. Make sure Ionic showcase server is running
2. Set server url in `Client.java`
3. Run app on multiple devices
4. `Add` `Delete` task should be reflected on both devices

### CODE REVIEW

File: app/src/main/java/com/m/androidNativeApp/MainActivity.java [hunk 1]
```diff
@@ -3,92 +3,222 @@
...
-                final int dataLength = mResponse.allTasks().size();
-
-                for (int i = 0; i < dataLength; i++) {
-                    taskTitle = mResponse.allTasks().get(i).fragments().taskFields().title();
-                    taskDescription = mResponse.allTasks().get(i).fragments().taskFields().description();
-                    taskId = mResponse.allTasks().get(i).fragments().taskFields().id();
-                    itemList.add(new Item(taskTitle, taskDescription, taskId));
+                for (Item item : itemList) {
+                    if (item.getId().equals(response.data().taskDeleted.fragments().taskFields.id())) {
+                        itemToRemove = item;
```
Comment (wei-lee) [inline app/src/main/java/com/m/androidNativeApp/MainActivity.java]: once the item is found, we should stop the loop.

File: app/src/main/java/com/m/androidNativeApp/MainActivity.java [hunk 2]
```diff
@@ -3,92 +3,222 @@
...
+                    }
                 }
+
+                itemList.remove(itemToRemove);
```
Comment (wei-lee) [inline app/src/main/java/com/m/androidNativeApp/MainActivity.java]: we should check that `itemToRemove` is actually not `null`.

File: app/src/main/java/com/m/androidNativeApp/MainActivity.java [hunk 3]
```diff
@@ -3,92 +3,222 @@
...
+    public void subscribeToAddTask() {
+        AddTaskSubscription addTaskSubscription = AddTaskSubscription
+                .builder()
+                .build();
+
+        apolloClient.subscribe(addTaskSubscription).execute(new ApolloSubscriptionCall.Callback<AddTaskSubscription.Data>() {
+            @Override
+            public void onResponse(@NotNull Response<AddTaskSubscription.Data> response) {
+                itemList.add(new Item(response.data().taskAdded.fragments().taskFields.title(), response.data().taskAdded.fragments().taskFields.description()
+                        , response.data().taskAdded.fragments().taskFields.id()));
```
Comment (wei-lee) [inline app/src/main/java/com/m/androidNativeApp/MainActivity.java]: is it possible that the `item` could be already existed in the `itemList`?

File: app/src/main/java/com/m/androidNativeApp/MainActivity.java [hunk 4]
```diff
@@ -3,92 +3,222 @@
...
 public class MainActivity extends AppCompatActivity {
 
-    private static final ApolloClient client = new Client().setupApollo();
-    //creating instance of Item Adapter Class with recycler view
     private String taskTitle, taskDescription, taskId;
     RecyclerView recyclerView;
-    ItemAdapter itemAdapter;
+    public ItemAdapter itemAdapter;
     List<Item> itemList;
+    Item itemToRemove;
```
Comment (wei-lee) [inline app/src/main/java/com/m/androidNativeApp/MainActivity.java]: does this variable need to be at class level? Or can we just use local variables in the methods where they are needed?

File: app/src/main/java/com/m/androidNativeApp/MainActivity.java [hunk 5]
```diff
@@ -3,92 +3,222 @@
...
                 .build();
 
-        client.mutate(deleteTask).enqueue(new ApolloCall.Callback<DeleteTaskMutation.Data>() {
+        apolloClient.mutate(deleteTask).enqueue(new ApolloCall.Callback<DeleteTaskMutation.Data>() {
+            @Override
+            public void onResponse(@NotNull final Response<DeleteTaskMutation.Data> response) {
+
+                for (Item item : itemList) {
+                    if (response.data() != null && item.getId().equals(response.data().deleteTask())) {
+                        itemToRemove = item;
```
Comment (wei-lee) [inline app/src/main/java/com/m/androidNativeApp/MainActivity.java]: again, we can stop the loop once the item is found

File: app/src/main/java/com/m/androidNativeApp/MainActivity.java [hunk 6]
```diff
@@ -3,92 +3,222 @@
...
             @Override
-            public void onResponse(@NotNull Response<DeleteTaskMutation.Data> response) {
-                System.out.println("Success");
+            public void onResponse(@NotNull Response<AllTasksQuery.Data> response) {
+                final AllTasksQuery.Data mResponse = response.data();
+
+                final int dataLength = mResponse.allTasks().size();
+
+                for (int i = 0; i < dataLength; i++) {
+                    taskTitle = mResponse.allTasks().get(i).fragments().taskFields().title();
```
Comment (wei-lee) [inline app/src/main/java/com/m/androidNativeApp/MainActivity.java]: `mResponse.allTasks().get(i)` is used many times here. Consider using a local variable here.

File: app/src/main/java/com/m/androidNativeApp/MainActivity.java (hunk 3 above)
Comment (MStokluska) [reply app/src/main/java/com/m/androidNativeApp/MainActivity.java]: Hi @wei-lee , please correct me if I'm wrong but I think that our itemList is populated in `get tasks` query and `add task` subscription, with `get tasks` being the initializer of itemList and subscription the "maintainer". If we would have a situation where `get tasks` would trigger aftear each `add task` mutation along with `add task` subscription then we could possibly have duplicated data. Our `get tasks` query is only triggered once, during app launch and any other time itemList is controlled by the subscriptions instead.

File: app/src/main/java/com/m/androidNativeApp/MainActivity.java (hunk 1 above)
Comment (MStokluska) [reply app/src/main/java/com/m/androidNativeApp/MainActivity.java]: great point, thank you. By doing so I can remove the `itemToRemove`

File: app/src/main/java/com/m/androidNativeApp/MainActivity.java (hunk 4 above)
Comment (MStokluska) [reply app/src/main/java/com/m/androidNativeApp/MainActivity.java]: Got this removed thanks to your other suggestion. Also made other attributes private.

**Review by wei-lee — APPROVED** [review]
needs rebase and it should be ready to merge.

**Pull request merged**